fall-2020/navigation.py

```diff
@@ -66,20 +66,53 @@ def __init__(self, GPS, steering):
 
     def set_desired_position(self, desired_position):
         """ Sets the desired position the TBM will attempt to move to.
-        
+
         Note: assume the user will pass in the desired_position parameter when using
         the interface 
         """
-        pass
+        self.desired_position = desired_position
+        return 0
 
     
     def update_current_position(self):
         """ Updates the current position of the TBM """
-        pass
+        self.current_position = self.GPS.getPos()
+        return 0
 
 
     def navigate(self):
-        #test pass
+        #tuple to store where to travel
+        x,y,z = self.desired_position-self.current_position
+
+        #looping through XYZ movement
+        if x > 0:
+            self.steering.move_forward()
+        else:
+            self.steering.stop()
+
+        if y == 0:
+            pass
+        elif y > 0:
+            self.steering.move_right(y)
+        elif y < 0:
+            y = abs(y)
+            self.steering.move_left(y)
```

Comment:
Could probably use else instead of the last elif, as it would save you from performing a comparison. If you are performing this function a few hundred times a second, it could potentially save a few clock cycles. 

Also, I like the explicit check for if y == 0, but if y is not > 0 and y is not < 0, then it has to be 0 and you could just do nothing in that case. 

The above comments also apply to z movement.

Reply:
Regarding the else - initially intended to use it! 

Completely forgot that python strictly doesn't allow comparison operands to be of different datatypes - was worried that the dynamically typed nature would allow for some odd condition where the variable XYZ was not a number and would pass the conditionals, hence activating the motor unintentionally. 

Also, regarding the y==0, I intended to use **break**, not **pass** - major brainfart right there. 

Logic behind the explicit check - if it's zero, it'll skip the checks for the other two (which might appear rather often than not since tunneling doesn't turn all that much, meaning either Y or Z is usually a zero)